compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -276,9 +280,9 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
                 let [pat_field] = &subpatterns[..] else { bug!() };
                 self.check_irrefutable(&pat_field.pattern, "`for` loop binding", None);
             } else {
-                non_exhaustive_match(
+                self.add_error_maybe(Err(non_exhaustive_match(
```

Comment:
```suggestion
                self.error = Err(non_exhaustive_match(
```
That should be enough. `ErrorGuaranteed` does not carry any meaningful information.